pennylane/_qubit_device.py

```diff
@@ -1484,6 +1490,84 @@ def var(self, observable, shot_range=None, bin_size=None):
         return np.squeeze(np.var(samples, axis=axis))
 
+    def _samples_to_counts(self, samples, obs, num_wires):
+        """Groups the samples into a dictionary showing number of occurences for
+        each possible outcome.
+
+        The format of the dictionary depends on obs.return_type, which is set when
+        calling measurements.counts by setting the kwarg all_outcomes (bool). Per default,
+        the dictionary will only contain the observed outcomes. Optionally (all_outcomes=True)
+        the dictionary will instead contain all possible outcomes, with a count of 0
+        for those not observed. See example.
+
+
+        Args:
+            samples: samples in an array of dimension ``(shots,len(wires))``
+            obs (Observable): the observable sampled
+            num_wires (int): number of wires the sampled observable was performed on
+
+        Returns:
+            dict: dictionary with format ``{'outcome': num_occurences}``, including all
+                outcomes for the sampled observable
+
+        **Example**
+
+            >>> samples
+            tensor([[0, 0],
+                    [0, 0],
+                    [1, 0]], requires_grad=True)
+
+            Per default, this will return:
+            >>> self._samples_to_counts(samples, obs, num_wires)
+            {'00': 2, '10': 1}
+
+            However, if obs.return_type is AllCounts, this will return:
+            >>> self._samples_to_counts(samples, obs, num_wires)
+            {'00': 2, '01': 0, '10': 1, '11': 0}
+
+            The variable all_outcomes can be set when running measurements.counts, i.e.:
+
+             .. code-block:: python3
+
+            dev = qml.device("default.qubit", wires=2, shots=4)
+
+            @qml.qnode(dev)
+            def circuit(x):
+                qml.RX(x, wires=0)
+                return qml.counts(all_outcomes=True)
+
+
+        """
+
+        outcomes = []
+
+        if isinstance(obs, MeasurementProcess):
+            # convert samples and outcomes (if using) from arrays to str for dict keys
+            samples = ["".join([str(s.item()) for s in sample]) for sample in samples]
+
+            if obs.return_type is AllCounts:
+                outcomes = self.generate_basis_states(num_wires)
+                outcomes = ["".join([str(o.item()) for o in outcome]) for outcome in outcomes]
+        else:
+            if obs.return_type is AllCounts:
+                try:
+                    outcomes = qml.eigvals(obs)
+                #  if observable has no info on eigenvalues, we cannot return this measurement
+                except qml.operation.EigvalsUndefinedError:
+                    warnings.warn(
+                        f"Warning: Error occurred in attempting to generate all outcomes."
+                        f"Cannot find outcomes for {obs.name} (EigvalsUndefinedError). "
+                        f"Only observed outcomes will be returned."
+                    )
```

Comment:
Add test to cover this.

Reply:
This was added to deal with an `EigenvalsUndefinedError` for `qml.Hermitian`, but it seems the error is no longer raised now that the eigenvalues are found using `outcomes = qml.eigvals(obs)` instead of `outcomes = obs.compute_eigvals()`. At least I haven't been able to find any instances now where this exception is relevant, so I've taken it out.